### ISSUE TITLE
Issues With install from source - Python

### ISSUE THREAD

**Issue by k6ctw** [coreemu/core#488]
While trying to build CORE from source on CentOS 7.8, and using the install.sh script it keeps failing saying:

checking for python3 script directory... ${prefix}/lib/python3.6/site-packages
checking for python3 extension module directory... ${exec_prefix}/lib/python3.6/site-packages
configure: error: please install python grpcio-tools

However, grpcio and grpcio-tools 1.3.x are both installed

Also, during the ./configure section of the script, it shows the following.

Requirement already satisfied: PyYAML==5.3 in /usr/local/lib64/python3.6/site-packages (from -r daemon/requirements.txt (line 18)) (5.3) Requirement already satisfied: six==1.14.0 in /usr/local/lib/python3.6/site-packages (from -r daemon/requirements.txt (line 19)) (1.14.0) Requirement already satisfied: setuptools in /usr/local/lib/python3.6/site-packages (from protobuf==3.11.3->-r daemon/requirements.txt (line 14)) (40.6.2)
ERROR: grpcio-tools 1.30.0 has requirement grpcio>=1.30.0, but you'll have grpcio 1.27.2 which is incompatible.
Installing collected packages: grpcio
  Attempting uninstall: grpcio
    Found existing installation: grpcio 1.30.0
    Uninstalling grpcio-1.30.0:
      Successfully uninstalled grpcio-1.30.0 Successfully installed grpcio-1.27.2 Bootstrapping the autoconf system...

**Comment by bharnden** [comment]
For CentOS you can try the following more manually, as install.sh was an attempt for convenience, but due to issues with system dependency conflicts like above, the end result is not ideal and why a shift to installing into a virtual environment, should at least help for that part.

Commands to try and ways to potentially mitigate your issues:
```
./bootstrap.sh
# the prefix option I believe is needed for CentOS pathing, maybe that has changed
./configure --prefix=/usr
make
sudo make install
```

The above is also assuming python3.6+ is getting picked up during configuration, if not you can force the python being targeted.
```
PYTHON=/path/to/python3.6 ./configure --prefix=/usr
```

Now that should at least get you into having things installed. If you are still getting an error about grpcio-tools, you can try to manually validate what is being checked.
```shell
python3 -m grpc_tools.protoc -h
```

If you can get grpcio-tools installed and to pass the above, it should work. All the dependencies in requirements.txt and their versions are a known working case, but likely older versions will work, unless really old. You can attempt to install them one by one until all are satisfied.

**Comment by bharnden** [comment]
Playing with CentOS myself lately this appears to be related to the dual installation of grpcio with different requirements. Which will hopefully be something resolved with #486.

Beyond that, there are some tweaks on the develop branch now to force the binary installation of grpcio as the user for building, which may solve your current specific problem.

**Comment by k6ctw** [comment]
Thaks for all the advice/info.  Hopefully I can get this working on Monday.  Have a GREAT rest of the weekend!73 & God Bless - Ken Miller, K6CTW-----Original Message-----

**Comment by k6ctw** [comment]
I was able to "fix" the issue of the conflicting re-install of the grpio tools, by changing the value in requirements.txt to the version that is actually installed.Now it's saying to install grpcio-tools however, those are already installed.  Oh well, on to the next.....-----Original Message-----

**Comment by bharnden** [comment]
Assuming you have only used pip from python3 (python3 -m pip). You can uninstall both grpcio related packages and validate that.

```shell
# uninstall
# may need to use sudo, if that was used for installation
python3 -m pip uninstall grpcio grpcio-tools

# verify grpcio packages are not present
# may need to use sudo, if that was used for installation
python3 -m pip freeze | grep grpcio

# repeat step 1 if multiple versions exist and are still being found

# re-install and force version to match runtime (12.7.2)
python3 -m pip install --user grpcio==12.7.2 grpcio-tools==12.7.2

# validate the command you expect exists, which is what configure uses for validation
python3 -m grpc_tools.protoc -h
```

**Comment by k6ctw** [comment]
The uninstall said they are not installed ?????So triend the install.  Got this failure:WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError("Can't connect to HTTPS URL because the SSL module is not available.",)': /simple/grpcio/Could not fetch URL https://pypi.org/simple/grpcio/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.org', port=443): Max retries exceeded with url: /simple/grpcio/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.",)) - skippingERROR: Could not find a version that satisfies the requirement grpcio==1.30.0 (from versions: none)ERROR: No matching distribution found for grpcio==1.30.0WARNING: pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.Could not fetch URL https://pypi.org/simple/pip/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.org', port=443): Max retries exceeded with url: /simple/pip/ (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available.",)) - skippingHave no idea what is missing   Thoughts?????-----Original Message-----

**Comment by k6ctw** [comment]
OK.  Thanks but that wasn't it.  Still getting the same issue with the SSLError.  not able to confirm SSL certificate.Note: Since this is a consistent issue, downloaded both of the grpcio packages:grpcio-1.30.0-cp36-cp36m-manylinux2010_x86_64.whlgrpcio_tools-1.30.0-cp36-cp36m-manylinux2010_x86_64.whlTried installing those using:python3 -m pip install --user /home/emane/Downloads/grpcio-1.30.0-cp36-cp36m-manylinux2010_x86_64.whl /home/emane/Downloads/grpcio_tools-1.30.0-cp36-cp36m-manylinux2010_x86_64.whl Fails for the same reason.  Also, not familiar with this python pip installation.  Thoughts?-----Original Message-----

**Comment by bharnden** [comment]
I mean you dont want 1.30.0 anyways, which is why I am saying 1.27.2.

Installation and SSL issues with pip seem more like a system/network issue at that point.

**Comment by k6ctw** [comment]
OK.  I'll go get the 1.27 versions and try installing again.  -----Original Message-----

**Comment by k6ctw** [comment]
Here is what's happening.  Its in the aclocal part...........Collecting grpcio-tools  Using cached grpcio_tools-1.30.0-cp36-cp36m-manylinux2010_x86_64.whl (2.5 MB)Collecting grpcio>=1.30.0  Using cached grpcio-1.30.0-cp36-cp36m-manylinux2010_x86_64.whl (3.0 MB)Requirement already satisfied: protobuf>=3.5.0.post1 in /usr/local/lib64/python3.6/site-packages (from grpcio-tools) (3.11.3)Requirement already satisfied: six>=1.5.2 in /usr/local/lib/python3.6/site-packages (from grpcio>=1.30.0->grpcio-tools) (1.14.0)Requirement already satisfied: setuptools in /usr/local/lib/python3.6/site-packages (from protobuf>=3.5.0.post1->grpcio-tools) (40.6.2)Installing collected packages: grpcio, grpcio-toolsSuccessfully installed grpcio-1.30.0 grpcio-tools-1.30.0However, these are supposedly not there............[root@centos7 protobuf-3.12.2.dist-info]# python3 -m pip uninstall grpcioWARNING: Skipping grpcio as it is not installed.[root@centos7 protobuf-3.12.2.dist-info]# python3 -m pip uninstall grpcio-toolsWARNING: Skipping grpcio-tools as it is not installed.[root@centos7 protobuf-3.12.2.dist-info]# -----Original Message-----

**Comment by bharnden** [comment]
The output is getting a bit garbled, probably from replying by email. But it still looks like it is trying to look for some cached version of 1.30.0 for some reason.

Maybe you need a clean environment. Hard to say without poking around myself.

**Comment by k6ctw** [comment]
OK - Got it to make and make install./bootstrap.sh                                                              (failed looking for grpcio_tools)so then ..........pip3 install --user grpcio_tools==1.27.2PYTHON=/usr/local/bin/python3 ./configure --prefix=/usrmakemake installNow I can start core-gui however, it giving me the following error:core-guiwarning: Tcl/Tk Img package not found  Thumbnails and other image types (JPG, PNG, etc.) will not be supported.  Please install it with:    yum install tkimg   (RedHat/Fedora)    sudo apt-get install libtk-img   (Debian/Ubuntu)Connecting to "core-daemon" (127.0.0.1:4038)...This feature requires a connection to the CORE daemon.Failed to connect to 127.0.0.1:4038!Have you started the CORE daemon with '/etc/init.d/core-daemon start'?  (Error: couldn't open socket: connection refused)So did that abd got this:/etc/init.d/core-daemon startstarting core-daemonTraceback (most recent call last):  File "/usr/bin/core-daemon", line 16, in <module>    from core import constantsModuleNotFoundError: No module named 'core'Very frustrating, so THANKS a bunch for all of your help and patience-----Original Message-----

**Comment by bharnden** [comment]
The tk warning for the gui, is due to an old dependency that newer versions of CentOS dont support and can probably be ignored.

The startup part again, is likely tied to issues with your python 
path. Since you see it complain about not finding "core".

You can try manually running the core-daemon, assuming you are seeing that on your PATH.
```
sudo core-daemon
```

Otherwise comparing where CORE was installed to, and what is on your python3 path and why those things are not aligning and fixing that issue.

```
# print out current python path
python3 -m site
```

probably the whole site-packages vs dist-packages thing, I would assume.

```
# this might be on python path
/usr/lib/python3.6/site-packages

# this  may be where core is getting installed, but not on python path, so it wont be found
/usr/lib/python3.6/dist-packages
```

**Comment by k6ctw** [comment]
Thanks again for all of the help.  Finally got it working by setting up a link between /root/.local and /usr/lib/python3.6  for the site-packages.  Have a great rest of the week!  73 ken-----Original Message-----

**Comment by bharnden** [comment]
this issue and some others, related to installation should hopefully be accounted for by the install changes available on develop as part of the change in #486 

One specific problem I found, related to what I can see you ran into is that the newer releases of grpc/grpcio-tools which due to older versions of pip typically found/installed on systems will force building from source, causing a very long build time and the need for build dependencies on the system.

CORE also relied on 1.27.2  as a runtime dependency, so that should all be aligned now on develop and will be kept in mind in the future.
